server/src/main/java/com/google/speech/tools/voxetta/servlets/UtteranceUploadServlet.java

```diff
@@ -12,80 +12,52 @@
-import com.google.appengine.api.blobstore.BlobKey;
-import com.google.appengine.api.blobstore.BlobstoreService;
-import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
-import com.google.appengine.api.datastore.DatastoreService;
-import com.google.appengine.api.datastore.DatastoreServiceFactory;
-import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.blobstore.BlobstoreFailureException;
+import com.google.appengine.api.datastore.DatastoreFailureException;
+import com.google.speech.tools.voxetta.data.Utterance; 
+import com.google.speech.tools.voxetta.services.DatastoreUtteranceService; 
 import java.io.IOException;
-import java.io.PrintWriter;
-import java.net.MalformedURLException;
-import java.net.URL;
-import java.util.List;
-import java.util.Map;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
  
-/** Servlet that handles Utterance Entities. */
+/** 
+ * Servlet that uploads Utterance Entities. 
+ */
 @WebServlet("/upload-utterance")
 public class UtteranceUploadServlet extends HttpServlet {
- 
-  private DatastoreService datastoreService = DatastoreServiceFactory.getDatastoreService();
-  private BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
- 
+
+  private DatastoreUtteranceService service = new DatastoreUtteranceService(); 
+  private String audio;
+
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
- 
-    // Get the URL of the audio file that has been uploaded to Blobstore
-    String audioKey = getUploadedFileKey(request, "audio");
-    
-    Entity utteranceEntity = new Entity("Utterance");
-    utteranceEntity.setProperty("audioKey", audioKey);
-    utteranceEntity.setProperty("userId", "FILLER");
-    utteranceEntity.setProperty("promptId", "FILLER");
-    utteranceEntity.setProperty("device", "FILLER");
-    utteranceEntity.setProperty("age", 100);
-    utteranceEntity.setProperty("gender", "FILLER");
-    
-    // Store the Utterance Entity in Datastore 
-    datastoreService.put(utteranceEntity);
- 
-    // Redirect back to the main recording page
-    response.sendRedirect("/index.html");
-  }
- 
-  // Returns the BlobKey of an uploaded file, or null if a file was not uploaded
-  private String getUploadedFileKey(HttpServletRequest request, String formInputElementName) {
-    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
-    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+    response.setContentType("application/json");
 
-    // Return null if an audio file was not uploaded
-    if (blobKeys == null || blobKeys.isEmpty()) {
-      return null; 
+    // Get the BlobKey of the audio file that has been uploaded to Blobstore
+    try {
+      audio = service.getAudio(request);
```

Comment:
Actually `IllegalStateException` is an unchecked exception, so we aren't syntactically required to catch anything here. I think we could remove this try/catch. Every checked `Exception` listed in the method's `throws` must be caught and handled, or added to the caller's `throws`. But unchecked exceptions don't need to be caught or listed in `throws`. It's actually a compile error to not handle a checked `Exception`.

Unchecked exceptions generally represent programming errors. For example, `NullPointerException` is an unchecked exception. Any code which references an object could raise this exception, but you don't see Java code littered with `catch (NullPointerException e)`. I believe the general guidance is you should try to catch unchecked exceptions if you think they would actually happen.

I know this is confusing. Error handling gets quite messy and makes code very ugly very quickly.

http://www.javapractices.com/topic/TopicAction.do?Id=129

Reply:
I was unfamiliar with checked vs unchecked error handling before, but your explanation and the linked document make complete sense. Removed try/catch